### PR TITLE
User: Refactor logout utility to use Redux

### PR DESCRIPTION
**Pull request by tyxla**
#### Changes proposed in this Pull Request

The last remaining legacy `lib/user/utils` function is `logout()`. This PR makes it a bit more independent by allowing it to accept an optional user prop and updates it to use Redux. It updates all locations to use that prop and removes the old function. Finally, we add a couple of tests to ensure the new functionality that uses `getLogoutUrl` without a user object still works.

Part of #24004 where we aim to reduxify `lib/user`.

#### Testing instructions

* Spin up Jetpack Cloud locally: `yarn run start-jetpack-cloud`
* Open `http://jetpack.cloud.localhost:3000/`
* Finish oauth if you haven't yet, pick a site.
* Click the profile button at the top right of your screen. 
* Click the link to logout.
* Verify you're redirected to Jetpack.com, and when you open `http://jetpack.cloud.localhost:3000/` again, it asks you to reauth.
* Start Calypso as usual with `yarn start` and continue with the next testing steps.
* Verify you are logged into WP.com.
* Go to http://calypso.localhost:3000/jetpack/connect/authorize?client_id=123123&site=example.com&nonce=nonce&secret=1&_wp_nonce=something&blogname=something&home_url=something&redirect_uri=something&scope=something&secret=something&site=something&site_url=something&state=something&woodna_service_name=login&debug
* Input your email address.
* On the next step, click "Create a new account"
* Verify it works like before and without errors in the console.
* Go through https://github.com/Automattic/wp-calypso/pull/53661#issuecomment-860828855 for setting up to test the desktop app dev version
* Verify logging out of the app still works the same way.
* Go to `/me/account` and close your account (best testable with a new fresh account without any upgrades) and verify it still works as it did before.
* Make sure you're logged in with a user that has 2FA enabled.
* Delete the `twostep_auth` cookie for `public-api.wordpress.com`.
* Go to `/me` and you'll see the reauthentication popup.
* Click the logout URL and verify it logs you out successfully, taking you to the WP.com landing page.
* Log in again and go to `/me`.
* Verify the log-out button in the sidebar works as it did before.
* Go through the invite flow and verify it still works well for both logged out and logged in users. Try also with a user that doesn't have a WP.com account yet.

### CODE REVIEW

File: client/lib/desktop-listeners/index.js [hunk 1]
```diff
@@ -138,7 +142,7 @@ const DesktopListeners = {
 	onSignout: function () {
 		debug( 'Signout' );
 
-		userUtilities.logout();
+		redirectToLogout( getCurrentUser( this.store.getState() ) );
```
Comment (jsnajdr) [inline client/lib/desktop-listeners/index.js]: I see some Redux patterns here:
- there could be a selector called `getLogoutUrl( state, redirect )` that reads from `state.currentUser.user` and constructs the logout URL.
- `redirectToLogout` could be an action thunk that is dispatched without any parameters. It reads all it needs from state internally. Is the Redux store accessible at all places where `redirectToLogout` is called?

File: client/lib/wpcom-xhr-wrapper/index.js [hunk 2]
```diff
@@ -20,7 +20,7 @@ export default async function ( params, callback ) {
 	return xhr( params, function ( error, response, headers ) {
 		if ( error && error.name === 'InvalidTokenError' ) {
 			debug( 'Invalid token error detected, authorisation probably revoked - logging out' );
-			userUtils.logout();
+			redirectToLogout();
```
Comment (jsnajdr) [inline client/lib/wpcom-xhr-wrapper/index.js]: This doesn't have access to `user.logout_URL` and therefore will always use the `config( 'logout_url' )` fallback without a nonce.

Comment (tyxla) [reply client/lib/wpcom-xhr-wrapper/index.js]: That's a fair point. Do you have any recommendations with regards to passing the user here? Using `redux-bridge` perhaps?

File: client/lib/desktop-listeners/index.js (hunk 1 above)
Comment (tyxla) [reply client/lib/desktop-listeners/index.js]: I was planning to go that way, but the instance in `wpcom-xhr-wrapper` was the one that doesn't have access to the store and we can't dispatch it there, not without using the redux bridge. What do you think?

**Comment by nsakaimbo** [comment]
👋 Thanks for keeping me in the loop on this. FYI, to test the changes with the "new" desktop app now, you can do the following:

1. Run calypso with `yarn start`
2. Run desktop with `cd desktop && yarn install --frozen-lockfile && yarn run dev:localhost`

The desktop workflow is now a lot faster and developer-friendly.

File: client/lib/desktop-listeners/index.js (hunk 1 above)
Comment (jsnajdr) [reply client/lib/desktop-listeners/index.js]: I think we can remove the `wpcom-xhr-wrapper` condition or replace it with a simple reload. Especially now when the Desktop app is just a wrapper and doesn't use OAuth. The condition was originally added in #612, solving a situation in Desktop app that doesn't have UI to reload the page.

The condition detects an expired OAuth token. By default, OAuth tokens expire in 14 days.

On the other hand, we don't have any check like this for cookies. The Calypso app will be getting failed REST responses until you reload the page. And the `wp_api` cookies in iframe proxy expire in only 2 days. The main `wordpress_sec` cookie lives practically forever, expiring in 3 years.

**Comment by tyxla** [comment]
Hey @jsnajdr thanks for the feedback. I've just made the `wpcom-xhr-wrapper` do a reload, and that allowed us to also convert `redirectToLogout` to a thunk, which I also took care of. This one could use another look when you get a chance.

File: client/components/jetpack/profile-dropdown/index.tsx [hunk 3]
```diff
@@ -39,7 +40,9 @@ const ProfileDropdown: React.FC = () => {
 
 	const trackedToggle = useTrackCallback( toggle, 'calypso_jetpack_masterbar_profile_toggle' );
 	const trackedClose = useTrackCallback( close, 'calypso_jetpack_masterbar_profile_close' );
-	const redirectToLogoutWithUser = useCallback( () => redirectToLogout( user ), [ user ] );
+	const redirectToLogoutWithUser = useCallback( () => dispatch( redirectToLogout() ), [
```
Comment (jsnajdr) [inline client/components/jetpack/profile-dropdown/index.tsx]: The `*WithUser` suffix no longer makes sense. A better name would be `redirectToLogout`, but that conflicts with the action creator identifier...

File: client/me/account-close/main.jsx [hunk 4]
```diff
@@ -49,7 +49,7 @@ class AccountSettingsClose extends Component {
 	UNSAFE_componentWillReceiveProps = ( nextProps ) => {
```
Comment (jsnajdr) [inline client/me/account-close/main.jsx]: This could be very straighforwardly updated to `componentDidUpdate( prevProps )`. Removing a deprecated method usage and moving the effect to the desired place.

File: client/me/reauth-required/index.jsx [hunk 5]
```diff
@@ -320,5 +316,8 @@ export default connect(
 		currentUser: getCurrentUser( state ),
```
Comment (jsnajdr) [inline client/me/reauth-required/index.jsx]: The `currentUser` prop is no longer used after this.

File: client/state/current-user/actions.js [hunk 6]
```diff
@@ -15,3 +18,15 @@ export function setCurrentUser( user ) {
 		user,
 	};
 }
+
+export async function redirectToLogout( postLogoutRedirectUrl ) {
```
Comment (jsnajdr) [inline client/state/current-user/actions.js]: This function can't be async, only the returned thunk can. Otherwise it's creating a promise of an action object instead of an action object, and that's going to fail:
```
redux.js:210 Uncaught Error: Actions must be plain objects. Use custom middleware for async actions.
```

Comment (jsnajdr) [reply client/state/current-user/actions.js]: This action could be in `state/logout/actions`. That module already contains a `logoutUser` action creator.

Which does the logout in a slightly different way: posting a REST XHR request to the `wp-login.php?action=logout-endpoint` instead of top-level redirecting to `wp-login.php?action=logout`.

I'm not sure what exactly are the pros and cons of each approach, and we should unify them eventually. The REST XHR approach is going to be troublesome with 3rd party cookies blocked though. It won't really log you out (i.e., remove the auth cookies) on a non-`wordpress.com` domain.

**Review by jsnajdr — COMMENTED** [review]
Found one major issue (action creator must not be async), everything else are nits 🙂

File: client/state/current-user/actions.js (hunk 6 above)
Comment (tyxla) [reply client/state/current-user/actions.js]: > This function can't be async, only the returned thunk can. 

I'm not sure how this happened, but thanks for catching it. Fixed in https://github.com/Automattic/wp-calypso/pull/53661/commits/3774400ac1c2252cdc7571e2fbcd0cb37ac523d2

Comment (tyxla) [reply client/state/current-user/actions.js]: > This action could be in `state/logout/actions`. That module already contains a `logoutUser` action creator.
> 
> Which does the logout in a slightly different way: posting a REST XHR request to the `wp-login.php?action=logout-endpoint` instead of top-level redirecting to `wp-login.php?action=logout`.
> 
> I'm not sure what exactly are the pros and cons of each approach, and we should unify them eventually. The REST XHR approach is going to be troublesome with 3rd party cookies blocked though. It won't really log you out (i.e., remove the auth cookies) on a non-`wordpress.com` domain.

Let's leave that unification for another time. These have been separate before and this PR's intention isn't to change that.

File: client/me/reauth-required/index.jsx (hunk 5 above)
Comment (tyxla) [reply client/me/reauth-required/index.jsx]: Of course. Removed in https://github.com/Automattic/wp-calypso/pull/53661/commits/fde89995cc649b0b571935233926eef07e17dacc

File: client/me/account-close/main.jsx (hunk 4 above)
Comment (tyxla) [reply client/me/account-close/main.jsx]: Good call. Updated in https://github.com/Automattic/wp-calypso/pull/53661/commits/2eee789dd25752bbfe0b0a0ea25e2d2220d731c7

File: client/components/jetpack/profile-dropdown/index.tsx (hunk 3 above)
Comment (tyxla) [reply client/components/jetpack/profile-dropdown/index.tsx]: Renamed to `redirectToLogoutUrl` in https://github.com/Automattic/wp-calypso/pull/53661/commits/0f0fd40690beaca03598ab3f4b3c61653abc7d57 which I think makes more sense 😉

**Comment by tyxla** [comment]
Thanks for the thorough review @jsnajdr! I've addressed the feedback, could you take another look when you get a chance?

File: client/me/account-close/main.jsx [hunk 7]
```diff
@@ -46,10 +46,10 @@ class AccountSettingsClose extends Component {
 		showSiteDropdown: true,
 	};
 
-	UNSAFE_componentWillReceiveProps = ( nextProps ) => {
+	componentDidUpdate = () => {
```
Comment (jsnajdr) [inline client/me/account-close/main.jsx]: nit: it should be a classic method on a prototype.

**Review by jsnajdr — APPROVED** [review]
Looks good :+1:

File: client/me/account-close/main.jsx (hunk 7 above)
Comment (tyxla) [reply client/me/account-close/main.jsx]: Ha, I wonder how it ended up being an arrow function in the first place. Fixing before 🚢 .

File: client/lib/wpcom-xhr-wrapper/index.js [hunk 8]
```diff
@@ -20,7 +15,7 @@ export default async function ( params, callback ) {
 	return xhr( params, function ( error, response, headers ) {
 		if ( error && error.name === 'InvalidTokenError' ) {
 			debug( 'Invalid token error detected, authorisation probably revoked - logging out' );
-			userUtils.logout();
+			window.location.reload();
```
Comment (jsnajdr) [inline client/lib/wpcom-xhr-wrapper/index.js:18]: @cbauerman Is there a reason why you might want to revert this change? Did it break anything in Jetpack Cloud? I noticed that you created a `revert/wpcom-xhr-bad-token-fallback` branch for that 🙂

Comment (tyxla) [reply client/lib/wpcom-xhr-wrapper/index.js:18]: Yeah, I was wondering the same thing, which is why I asked here: p1624086030240200/1624047611.234600-slack-CQXNTE9K9 . I'm keeping an eye on this and will revisit once we have more information.
